// @visactor/vtable/cjs/index.js
"use strict";

var __createBinding = this && this.__createBinding || (Object.create ? function(o, m, k, k2) {
    void 0 === k2 && (k2 = k);
    var desc = Object.getOwnPropertyDescriptor(m, k);
    desc && !("get" in desc ? !m.__esModule : desc.writable || desc.configurable) || (desc = {
        enumerable: !0,
        get: function() {
            return m[k];
        }
    }), Object.defineProperty(o, k2, desc);
} : function(o, m, k, k2) {
    void 0 === k2 && (k2 = k), o[k2] = m[k];
}), __setModuleDefault = this && this.__setModuleDefault || (Object.create ? function(o, v) {
    Object.defineProperty(o, "default", {
        enumerable: !0,
        value: v
    });
} : function(o, v) {
    o.default = v;
}), __importStar = this && this.__importStar || function(mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (null != mod) for (var k in mod) "default" !== k && Object.prototype.hasOwnProperty.call(mod, k) && __createBinding(result, mod, k);
    return __setModuleDefault(result, mod), result;
}, __exportStar = this && this.__exportStar || function(m, exports) {
    for (var p in m) "default" === p || Object.prototype.hasOwnProperty.call(exports, p) || __createBinding(exports, m, p);
};

Object.defineProperty(exports, "__esModule", {
    value: !0
}), exports.EventTarget = exports.PIVOT_TABLE_EVENT_TYPE = exports.PIVOT_CHART_EVENT_TYPE = exports.TABLE_EVENT_TYPE = exports.plugins = exports.restoreMeasureText = exports.setCustomAlphabetCharSet = exports.graphicUtil = exports.renderChart = exports.updateCell = exports.CustomLayout = exports.DataStatistics = exports.register = exports.clearGlobal = exports.getIcons = exports.data = exports.themes = exports.textMeasure = exports.PivotChart = exports.PivotTableSimple = exports.PivotTable = exports.ListTableSimple = exports.ListTable = exports.core = exports.TYPES = exports.version = exports.Icon = exports.getTargetCell = exports.getDataCellPath = void 0;

const vrender_1 = require("./vrender");

Object.defineProperty(exports, "graphicUtil", {
    enumerable: !0,
    get: function() {
        return vrender_1.graphicUtil;
    }
}), (0, vrender_1.registerForVrender)();

const TYPES = __importStar(require("./ts-types"));

exports.TYPES = TYPES;

const core = __importStar(require("./core"));

exports.core = core;

const data = __importStar(require("./data"));

exports.data = data;

const icons = __importStar(require("./icons")), register = __importStar(require("./register"));

exports.register = register;

const themes = __importStar(require("./themes"));

exports.themes = themes;

const DataStatistics = __importStar(require("./dataset/DataStatistics"));

exports.DataStatistics = DataStatistics;

const plugins = __importStar(require("./plugins"));

exports.plugins = plugins;

const ListTable_all_1 = require("./ListTable-all");

Object.defineProperty(exports, "ListTable", {
    enumerable: !0,
    get: function() {
        return ListTable_all_1.ListTableAll;
    }
});

const ListTable_simple_1 = require("./ListTable-simple");

Object.defineProperty(exports, "ListTableSimple", {
    enumerable: !0,
    get: function() {
        return ListTable_simple_1.ListTableSimple;
    }
});

const PivotTable_all_1 = require("./PivotTable-all");

Object.defineProperty(exports, "PivotTable", {
    enumerable: !0,
    get: function() {
        return PivotTable_all_1.PivotTableAll;
    }
});

const PivotTable_simple_1 = require("./PivotTable-simple");

Object.defineProperty(exports, "PivotTableSimple", {
    enumerable: !0,
    get: function() {
        return PivotTable_simple_1.PivotTableSimple;
    }
});

const PivotChart_1 = require("./PivotChart");

Object.defineProperty(exports, "PivotChart", {
    enumerable: !0,
    get: function() {
        return PivotChart_1.PivotChart;
    }
});

const CustomLayout = __importStar(require("./render/layout"));

exports.CustomLayout = CustomLayout;

const cell_helper_1 = require("./scenegraph/group-creater/cell-helper");

Object.defineProperty(exports, "updateCell", {
    enumerable: !0,
    get: function() {
        return cell_helper_1.updateCell;
    }
});

const chart_render_helper_1 = require("./scenegraph/graphic/contributions/chart-render-helper");

Object.defineProperty(exports, "renderChart", {
    enumerable: !0,
    get: function() {
        return chart_render_helper_1.renderChart;
    }
});

const text_measure_1 = require("./scenegraph/utils/text-measure");

Object.defineProperty(exports, "restoreMeasureText", {
    enumerable: !0,
    get: function() {
        return text_measure_1.restoreMeasureText;
    }
}), Object.defineProperty(exports, "setCustomAlphabetCharSet", {
    enumerable: !0,
    get: function() {
        return text_measure_1.setCustomAlphabetCharSet;
    }
}), Object.defineProperty(exports, "textMeasure", {
    enumerable: !0,
    get: function() {
        return text_measure_1.textMeasure;
    }
});

var get_data_path_1 = require("./tools/get-data-path");

Object.defineProperty(exports, "getDataCellPath", {
    enumerable: !0,
    get: function() {
        return get_data_path_1.getDataCellPath;
    }
}), __exportStar(require("./render/jsx"), exports);

var util_1 = require("./event/util");

Object.defineProperty(exports, "getTargetCell", {
    enumerable: !0,
    get: function() {
        return util_1.getTargetCell;
    }
});

var icon_1 = require("./scenegraph/graphic/icon");

function getIcons() {
    return icons.get();
}

function clearGlobal() {
    register.clearAll();
}

Object.defineProperty(exports, "Icon", {
    enumerable: !0,
    get: function() {
        return icon_1.Icon;
    }
}), exports.version = "1.18.3", exports.getIcons = getIcons, exports.clearGlobal = clearGlobal, 
TYPES.AggregationType, __exportStar(require("./components"), exports), __exportStar(require("./scenegraph/group-creater/cell-type"), exports);

var TABLE_EVENT_TYPE_1 = require("./core/TABLE_EVENT_TYPE");

Object.defineProperty(exports, "TABLE_EVENT_TYPE", {
    enumerable: !0,
    get: function() {
        return TABLE_EVENT_TYPE_1.TABLE_EVENT_TYPE;
    }
});

var PIVOT_TABLE_EVENT_TYPE_1 = require("./ts-types/pivot-table/PIVOT_TABLE_EVENT_TYPE");

Object.defineProperty(exports, "PIVOT_CHART_EVENT_TYPE", {
    enumerable: !0,
    get: function() {
        return PIVOT_TABLE_EVENT_TYPE_1.PIVOT_CHART_EVENT_TYPE;
    }
}), Object.defineProperty(exports, "PIVOT_TABLE_EVENT_TYPE", {
    enumerable: !0,
    get: function() {
        return PIVOT_TABLE_EVENT_TYPE_1.PIVOT_TABLE_EVENT_TYPE;
    }
});

var EventTarget_1 = require("./event/EventTarget");

Object.defineProperty(exports, "EventTarget", {
    enumerable: !0,
    get: function() {
        return EventTarget_1.EventTarget;
    }
});
