// @visactor/vtable/es/index.js
import { graphicUtil, registerForVrender } from "./vrender";

registerForVrender();

import * as TYPES from "./ts-types";

import * as core from "./core";

import * as data from "./data";

import * as icons from "./icons";

import * as register from "./register";

import * as themes from "./themes";

import * as DataStatistics from "./dataset/DataStatistics";

import * as plugins from "./plugins";

import { ListTableAll as ListTable } from "./ListTable-all";

import { ListTableSimple } from "./ListTable-simple";

import { PivotTableAll as PivotTable } from "./PivotTable-all";

import { PivotTableSimple } from "./PivotTable-simple";

import { PivotChart } from "./PivotChart";

import * as CustomLayout from "./render/layout";

import { updateCell } from "./scenegraph/group-creater/cell-helper";

import { renderChart } from "./scenegraph/graphic/contributions/chart-render-helper";

import { restoreMeasureText, setCustomAlphabetCharSet, textMeasure } from "./scenegraph/utils/text-measure";

export { getDataCellPath } from "./tools/get-data-path";

export * from "./render/jsx";

export { getTargetCell } from "./event/util";

export { Icon } from "./scenegraph/graphic/icon";

export const version = "1.18.3";

export { TYPES, core, ListTable, ListTableSimple, PivotTable, PivotTableSimple, PivotChart, textMeasure, themes, data, getIcons, clearGlobal, register, DataStatistics, CustomLayout, updateCell, renderChart, graphicUtil, setCustomAlphabetCharSet, restoreMeasureText, plugins };

function getIcons() {
    return icons.get();
}

function clearGlobal() {
    register.clearAll();
}

TYPES.AggregationType;

export * from "./components";

export * from "./scenegraph/group-creater/cell-type";

export { TABLE_EVENT_TYPE } from "./core/TABLE_EVENT_TYPE";

export { PIVOT_CHART_EVENT_TYPE, PIVOT_TABLE_EVENT_TYPE } from "./ts-types/pivot-table/PIVOT_TABLE_EVENT_TYPE";

export { EventTarget } from "./event/EventTarget";
