// @visactor/vtable/cjs/scenegraph/group-creater/cell-type/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
    value: !0
}), exports.registerVideoCell = exports.registerTextCell = exports.registerSparkLineCell = exports.registerButtonCell = exports.registerSwitchCell = exports.registerRadioCell = exports.registerProgressBarCell = exports.registerImageCell = exports.registerCheckboxCell = exports.registerChartCell = void 0;

const factory_1 = require("../../../core/factory"), chart_cell_1 = require("./chart-cell"), checkbox_cell_1 = require("./checkbox-cell"), image_cell_1 = require("./image-cell"), radio_cell_1 = require("./radio-cell"), switch_cell_1 = require("./switch-cell"), spark_line_cell_1 = require("./spark-line-cell"), video_cell_1 = require("./video-cell"), text_cell_1 = require("./text-cell"), progress_bar_cell_1 = require("./progress-bar-cell"), get_axis_domain_1 = require("../../../layout/chart-helper/get-axis-domain"), button_cell_1 = require("./button-cell"), registerChartCell = () => {
    factory_1.Factory.registerFunction("createChartCellGroup", chart_cell_1.createChartCellGroup), 
    factory_1.Factory.registerFunction("getAxisDomainRangeAndLabels", get_axis_domain_1.getAxisDomainRangeAndLabels);
};

exports.registerChartCell = registerChartCell;

const registerCheckboxCell = () => {
    factory_1.Factory.registerFunction("createCheckboxCellGroup", checkbox_cell_1.createCheckboxCellGroup);
};

exports.registerCheckboxCell = registerCheckboxCell;

const registerImageCell = () => {
    factory_1.Factory.registerFunction("createImageCellGroup", image_cell_1.createImageCellGroup);
};

exports.registerImageCell = registerImageCell;

const registerProgressBarCell = () => {
    factory_1.Factory.registerFunction("createProgressBarCell", progress_bar_cell_1.createProgressBarCell);
};

exports.registerProgressBarCell = registerProgressBarCell;

const registerRadioCell = () => {
    factory_1.Factory.registerFunction("createRadioCellGroup", radio_cell_1.createRadioCellGroup);
};

exports.registerRadioCell = registerRadioCell;

const registerSwitchCell = () => {
    factory_1.Factory.registerFunction("createSwitchCellGroup", switch_cell_1.createSwitchCellGroup);
};

exports.registerSwitchCell = registerSwitchCell;

const registerButtonCell = () => {
    factory_1.Factory.registerFunction("createButtonCellGroup", button_cell_1.createButtonCellGroup);
};

exports.registerButtonCell = registerButtonCell;

const registerSparkLineCell = () => {
    factory_1.Factory.registerFunction("createSparkLineCellGroup", spark_line_cell_1.createSparkLineCellGroup);
};

exports.registerSparkLineCell = registerSparkLineCell;

const registerTextCell = () => {
    factory_1.Factory.registerFunction("createTextCellGroup", text_cell_1.createCellGroup);
};

exports.registerTextCell = registerTextCell;

const registerVideoCell = () => {
    factory_1.Factory.registerFunction("createVideoCellGroup", video_cell_1.createVideoCellGroup);
};

exports.registerVideoCell = registerVideoCell;
