// @visactor/vtable/cjs/state/select/custom-select.js
"use strict";

Object.defineProperty(exports, "__esModule", {
    value: !0
}), exports.addCustomSelectRanges = exports.deletaCustomSelectRanges = void 0;

const vrender_1 = require("./../../vrender"), update_select_border_1 = require("../../scenegraph/select/update-select-border");

function deletaCustomSelectRanges(state) {
    const {customSelectedRangeComponents: customSelectedRangeComponents} = state.table.scenegraph;
    customSelectedRangeComponents.forEach(((selectComp, key) => {
        selectComp.rect.delete();
    })), customSelectedRangeComponents.clear(), state.select.customSelectRanges = [];
}

function addCustomSelectRanges(customSelectRanges, state) {
    const {customSelectedRangeComponents: customSelectedRangeComponents} = state.table.scenegraph;
    customSelectRanges.forEach((customRange => {
        var _a, _b, _c, _d;
        const {range: range, style: style} = customRange, rect = (0, vrender_1.createRect)({
            fill: null !== (_a = style.cellBgColor) && void 0 !== _a && _a,
            stroke: null !== (_b = style.cellBorderColor) && void 0 !== _b && _b,
            lineWidth: null !== (_c = style.cellBorderLineWidth) && void 0 !== _c ? _c : 0,
            lineDash: null !== (_d = style.cellBorderLineDash) && void 0 !== _d ? _d : [],
            pickable: !1
        });
        customSelectedRangeComponents.set(`${range.start.col}-${range.start.row}-${range.end.col}-${range.end.row}`, {
            rect: rect,
            role: "body"
        });
    })), state.select.customSelectRanges = customSelectRanges, (0, update_select_border_1.updateAllSelectComponent)(state.table.scenegraph), 
    state.table.scenegraph.updateNextFrame();
}

exports.deletaCustomSelectRanges = deletaCustomSelectRanges, exports.addCustomSelectRanges = addCustomSelectRanges;
