// @visactor/vtable/cjs/render/jsx/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
    value: !0
}), exports.VRadio = exports.VCheckBox = exports.VTag = exports.jsx = exports.VText = exports.VPolygon = exports.VPath = exports.VLine = exports.VImage = exports.VGlyph = exports.VCircle = exports.VArea = exports.VArc = exports.VRect = exports.VSymbol = exports.VGroup = void 0;

var vrender_1 = require("./../../vrender");

Object.defineProperty(exports, "VGroup", {
    enumerable: !0,
    get: function() {
        return vrender_1.VGroup;
    }
}), Object.defineProperty(exports, "VSymbol", {
    enumerable: !0,
    get: function() {
        return vrender_1.VSymbol;
    }
}), Object.defineProperty(exports, "VRect", {
    enumerable: !0,
    get: function() {
        return vrender_1.VRect;
    }
}), Object.defineProperty(exports, "VArc", {
    enumerable: !0,
    get: function() {
        return vrender_1.VArc;
    }
}), Object.defineProperty(exports, "VArea", {
    enumerable: !0,
    get: function() {
        return vrender_1.VArea;
    }
}), Object.defineProperty(exports, "VCircle", {
    enumerable: !0,
    get: function() {
        return vrender_1.VCircle;
    }
}), Object.defineProperty(exports, "VGlyph", {
    enumerable: !0,
    get: function() {
        return vrender_1.VGlyph;
    }
}), Object.defineProperty(exports, "VImage", {
    enumerable: !0,
    get: function() {
        return vrender_1.VImage;
    }
}), Object.defineProperty(exports, "VLine", {
    enumerable: !0,
    get: function() {
        return vrender_1.VLine;
    }
}), Object.defineProperty(exports, "VPath", {
    enumerable: !0,
    get: function() {
        return vrender_1.VPath;
    }
}), Object.defineProperty(exports, "VPolygon", {
    enumerable: !0,
    get: function() {
        return vrender_1.VPolygon;
    }
}), Object.defineProperty(exports, "VText", {
    enumerable: !0,
    get: function() {
        return vrender_1.VText;
    }
}), Object.defineProperty(exports, "jsx", {
    enumerable: !0,
    get: function() {
        return vrender_1.jsx;
    }
});

const vrender_2 = require("./../../vrender");

function VTag(params) {
    return new vrender_2.Tag(params ? params.attribute : {});
}

function VCheckBox(params) {
    return new vrender_2.CheckBox(params ? params.attribute : {});
}

function VRadio(params) {
    return new vrender_2.Radio(params ? params.attribute : {});
}

exports.VTag = VTag, exports.VCheckBox = VCheckBox, exports.VRadio = VRadio;
