// @visactor/vtable/cjs/components/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
    value: !0
}), exports.registerTooltip = exports.registerTitle = exports.registerMenu = exports.registerLegend = exports.registerEmptyTip = exports.registerAxis = void 0;

const factory_1 = require("../core/factory"), get_axis_config_1 = require("../layout/chart-helper/get-axis-config"), axis_1 = require("./axis/axis"), get_axis_component_size_1 = require("./axis/get-axis-component-size"), empty_tip_1 = require("./empty-tip/empty-tip"), create_legend_1 = require("./legend/create-legend"), MenuHandler_1 = require("./menu/dom/MenuHandler"), title_1 = require("./title/title"), TooltipHandler_1 = require("./tooltip/TooltipHandler"), registerAxis = () => {
    factory_1.Factory.registerComponent("axis", axis_1.CartesianAxis), factory_1.Factory.registerFunction("computeAxisComponentWidth", get_axis_component_size_1.computeAxisComponentWidth), 
    factory_1.Factory.registerFunction("computeAxisComponentHeight", get_axis_component_size_1.computeAxisComponentHeight), 
    factory_1.Factory.registerFunction("getAxisConfigInPivotChart", get_axis_config_1.getAxisConfigInPivotChart);
};

exports.registerAxis = registerAxis;

const registerEmptyTip = () => {
    factory_1.Factory.registerComponent("emptyTip", empty_tip_1.EmptyTip);
};

exports.registerEmptyTip = registerEmptyTip;

const registerLegend = () => {
    factory_1.Factory.registerFunction("createLegend", create_legend_1.createLegend);
};

exports.registerLegend = registerLegend;

const registerMenu = () => {
    factory_1.Factory.registerComponent("menuHandler", MenuHandler_1.MenuHandler);
};

exports.registerMenu = registerMenu;

const registerTitle = () => {
    factory_1.Factory.registerComponent("title", title_1.Title);
};

exports.registerTitle = registerTitle;

const registerTooltip = () => {
    factory_1.Factory.registerComponent("tooltipHandler", TooltipHandler_1.TooltipHandler);
};

exports.registerTooltip = registerTooltip;
