// @visactor/vtable/cjs/tools/get-data-path/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
    value: !0
}), exports.getDataCellPath = void 0;

const vutils_1 = require("@visactor/vutils"), create_dataset_1 = require("./create-dataset"), pivot_header_layout_1 = require("../../layout/pivot-header-layout"), tree_helper_1 = require("../../layout/tree-helper"), global_1 = require("../global");

function getDataCellPath(options, data, compareFunc) {
    var _a, _b;
    const results = (0, create_dataset_1.createDataset)(options);
    let columnDimensionTree = results.columnDimensionTree, rowDimensionTree = results.rowDimensionTree, isNeedResetColumnDimensionTree = !1, isNeedResetRowDimensionTree = !1;
    options.columnTree && options.indicatorsAsCol && !columnDimensionTree.dimensionKeys.contain(global_1.IndicatorDimensionKeyPlaceholder) && (isNeedResetColumnDimensionTree = !0), 
    options.rowTree && (options.indicatorsAsCol || rowDimensionTree.dimensionKeys.contain(global_1.IndicatorDimensionKeyPlaceholder) || (isNeedResetRowDimensionTree = !0));
    const {dataset: dataset, layoutNodeId: layoutNodeId} = results, mockTable = {
        options: options,
        layoutNodeId: layoutNodeId,
        internalProps: options,
        isPivotChart: () => !0,
        pivotChartAxes: [],
        _selectedDataItemsInChart: [],
        _getActiveChartInstance: () => ({
            updateState: () => {}
        })
    };
    options.columnTree && !isNeedResetColumnDimensionTree || (columnDimensionTree = new tree_helper_1.DimensionTree(null !== (_a = dataset.colHeaderTree) && void 0 !== _a ? _a : [], layoutNodeId)), 
    options.rowTree && !isNeedResetRowDimensionTree || (rowDimensionTree = new tree_helper_1.DimensionTree(null !== (_b = dataset.rowHeaderTree) && void 0 !== _b ? _b : [], layoutNodeId));
    const layoutMap = new pivot_header_layout_1.PivotHeaderLayoutMap(mockTable, dataset, columnDimensionTree, rowDimensionTree);
    for (let col = 0; col < layoutMap.colCount; col++) for (let row = 0; row < layoutMap.rowCount; row++) {
        if (layoutMap.isHeader(col, row)) continue;
        const cellDimensionPath = layoutMap.getCellHeaderPaths(col, row), colKeys = cellDimensionPath.colHeaderPaths.map((colPath => {
            var _a;
            return null !== (_a = colPath.indicatorKey) && void 0 !== _a ? _a : colPath.value;
        })), rowKeys = cellDimensionPath.rowHeaderPaths.map((rowPath => {
            var _a;
            return null !== (_a = rowPath.indicatorKey) && void 0 !== _a ? _a : rowPath.value;
        })), aggregator = dataset.getAggregator(layoutMap.indicatorsAsCol ? rowKeys : rowKeys.slice(0, -1), layoutMap.indicatorsAsCol ? colKeys.slice(0, -1) : colKeys, layoutMap.getIndicatorKey(col, row)), result = compareData(aggregator.value ? aggregator.value() : void 0, data, col, row, layoutMap, compareFunc);
        if (result) return result;
    }
}

function compareData(data1, data2, col, row, layoutMap, compareFunc) {
    if ((0, vutils_1.isArray)(data1)) for (let i = 0; i < data1.length; i++) if (compareFunc ? compareFunc(data1[i], data2) : defaultCompare(data1[i], data2)) return layoutMap.getCellHeaderPaths(col, row);
}

function defaultCompare(a, b) {
    return JSON.stringify(a) === JSON.stringify(b);
}

exports.getDataCellPath = getDataCellPath;
